File: 第2次课/第2题/2 拉格朗日插值/2 切比雪夫高斯罗巴托点/lagrangeinterpolation.py

import numpy as np
# import pandas as pd
def lagrangeinterpolation(xy,x):
    n=len(xy[0]);

    #获取定点集x,y
    x0=xy[0];
    y0=xy[1];
    #待求集长度
    m=len(x);
    #求l[]
    l=np.zeros((m,n));
    for i in range(0,m):
        #指定x，指定k，计算l
        tempx=x[i];
        for k in range(0,n):
            templ=1;
            for j in range(0,n):
                if(j!=k):
                    templ*=((tempx-x0[j])/(x0[k]-x0[j]));#这里是x0
            l[i,k]=templ;

    y0=np.matrix(y0);#word，problem
    # l=np.matrix(l);
    # x=np.matrix(x);
    y=l*y0.T;#function 这次都是行向量,最后结果是矩阵，mulsum出错
    y=np.asarray(y);#需要
    y=np.squeeze(y)#需要接收
    return y

File: 第2次课/第1题/2 拉格朗日插值/1 等分点/lagrangeinterpolation.m
function y = lagrangeinterpolation(x0,y0,x )
n=length(x0);%获取长度
m=length(x);

L0=zeros(n,1);%计算l值，初始化l0是n，1

for k=1:n
     L=1;
     for j=1:n
         if j~=k%不等于下的两个乘积
             L=L/(x0(k)-x0(j));%下半部分
         end
     end
     L0(k)=L;
end


for l=1:m
   z=x(l);
   s=0;
   for k=1:n
      L=L0(k);
      for j=1:n
         if j~=k
            L=L*(z-x0(j));%上半部分
         end
      end
      s=s+L*y0(k);
   end
   y(l)=s;
end


end

